gather-scheduler.js

```diff
@@ -18,99 +18,124 @@
+        return resolve(driver.beginEmulation());
+      }
+
+      // noop if no mobile emulation
+      resolve();
+    }).then(_ => {
+      return driver.cleanAndDisableBrowserCaches();
+    }).then(_ => {
       // Force SWs to update on load.
-      .then(_ => driver.forceUpdateServiceWorkers())
-
-      // Gather: setup phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.setup(options)))
-
-      // Enable tracing and network record collection.
-      .then(_ => driver.beginTrace())
-      .then(_ => driver.beginNetworkCollect())
-
-      // Gather: beforePageLoad phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.beforePageLoad(options)))
-
-      // Load the page (if the CLI / extension want it loaded).
-      .then(_ => {
-        if (loadPage) {
-          return driver.gotoURL(url, driver.WAIT_FOR_LOADED);
-        }
-
-        return Promise.resolve();
-      })
-
-      // Gather: afterPageLoad phase
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterPageLoad(options)))
+      return driver.forceUpdateServiceWorkers();
+    });
+  }
 
-      // Disable network collection; grab records.
-      .then(_ => driver.endNetworkCollect())
-      .then(networkRecords => {
-        tracingData.networkRecords = networkRecords;
-      })
+  // Enable tracing and network record collection.
+  static beginPassiveCollection(driver) {
+    return driver.beginTrace().then(_ => {
+      return driver.beginNetworkCollect();
+    });
+  }
 
-      // Disable tracing; grab records.
-      .then(_ => driver.endTrace())
-      .then(traceContents => {
-        tracingData.traceContents = traceContents;
-      })
+  static endPassiveCollection(driver, tracingData) {
+    return driver.endNetworkCollect().then(networkRecords => {
+      tracingData.networkRecords = networkRecords;
+    }).then(_ => {
+      return driver.endTrace();
+    }).then(traceContents => {
+      tracingData.traceContents = traceContents;
+    });
+  }
 
-      // Gather: afterTraceCollected phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterTraceCollected(options, tracingData)))
+  static _runPhase(gatherers, gatherFun) {
+    return gatherers.reduce((chain, gatherer) => {
+      return chain.then(_ => gatherFun(gatherer));
+    }, Promise.resolve());
+  }
 
-      // Disconnect the driver.
-      .then(_ => driver.disconnect())
+  static run(gatherers, options) {
+    const driver = options.driver;
+    const tracingData = {};
 
-      // Gather: tearDown phase.
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
+      throw new Error('You must provide a url to scheduler');
+    }
 
+    return driver.connect().then(_ => {
+      return GatherScheduler.setupDriver(driver, gatherers, options);
```

Comment:
that's why I was saying make the methods just functions (since they're already operating as ones). Then you can just do, like,
```
    return driver.connect()
      .then(_ => setupDriver(driver, gatherers, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.setup(options)))
      .then(_ => beginPassiveCollection(driver))
      .then(_ => runPhase(gatherers, gatherer => gatherer.beforePageLoad(options)))
      .then(_ => loadPage(driver, gatherers, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.afterPageLoad(options)))
      .then(_ => endPassiveCollection(driver, tracingData))
      .then(_ => runPhase(gatherers,
          gatherer => gatherer.afterTraceCollected(options, tracingData)))
      .then(_ => runPhase(gatherers, gatherer => gatherer.reloadSetup(options)))
      .then(_ => runPhase(gatherers, gatherer => gatherer.beforeReloadPageLoad(options)))
      .then(_ => reloadPage(driver, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.afterReloadPageLoad(options)))
      .then(_ => driver.disconnect())
      .then(_ => runPhase(gatherers, gatherer => gatherer.tearDown(options)))
      .then(_ => {
        // Collate all the gatherer results.
        return gatherers.map(g => g.artifact).concat(
            {networkRecords: tracingData.networkRecords},
            {traceContents: tracingData.traceContents});
      });
```